### DISCUSSION TITLE
MQTT/TCP works fine on a linux OS but not on Windows

### DISCUSSION THREAD

**Discussion by BrunoSilvaUM** [cesanta/mongoose#2286] [Ask questions, get answers] [answered]
Hi,
I'm using an STM32F767 an I've been trying to use your library to work with a mosquitto broker that is hosted on a windows PC.
I've got it to work perfectly fine on linux but never on windows.
First of all, I can only get an IP on Windows when I declare it staticaly, never worked with DHCP. On linux it works fine.
When IP is static I can ping it from the host.
Im using FreeRTOS CMSIS2 ARCH (have tried with CMSISv1 too) and I'm using mongoose TCPIP Stack (Also have tried with LWIP stack and didnt work).
Firewall is disabled on windows. GPIO ETH Pins are correctly assigned. I've also tried many configurations on the Network and share center Ethernet properties, like enabling ICS. 
Is there any windows setting or compiler flag that I have missed? I've tried to follow all the STM32 tutorials on this matter.

Also tried connecting to HiveMQ public broker and it gives me **this output:**
(On linux it works fine)

```
...
d444   2 main.c:178:fn                  12 CLOSED

df39   2 main.c:129:fn                  13 CREATED

eaf6   1 event.c:21:mg_error            13 0xffffffff DNS timeout

eafc   1 main.c:134:fn                  13 ERROR DNS timeout

ebca   2 main.c:178:fn                  13 CLOSED

//(LOOPS FOREVER LIKE THIS)
....

```

Here's an image of the output i get on Windows as well as the flags i use:

**output on Windows:** When running the code bellow with a static IP defined and trying to connect to a mosquitto broker
![si1](https://github.com/cesanta/mongoose/assets/62550297/08b8acfa-73a4-43c3-a68a-4d35ff8f705c)

**Output on Linux (Works fine):**
![MQTT Test](https://github.com/cesanta/mongoose/assets/62550297/dd627f40-24c7-445a-b82c-983028dc014c)

**Flags:** 
![image](https://github.com/cesanta/mongoose/assets/62550297/58939c7a-2f2b-4fe3-a273-d27d82f3da8e)

![image](https://github.com/cesanta/mongoose/assets/62550297/a2856089-5cc0-4a5e-9066-d64e9096aae1)

**Mosquitto settings are:**
```
listener 1883
allow_anonymous true
```

**mongoose_custom.h:**

```
#ifndef __MONGOOSE_CUSTOM_H__
#define __MONGOOSE_CUSTOM_H__

#if defined(MG_ARCH)
#define MG_STMPACK_ARCH MG_ARCH
#undef MG_ARCH
#elif defined(__GNUC__)
#define MG_STMPACK_ARCH MG_ARCH_NEWLIB
#endif

#define MG_ENABLE_PACKED_FS  1

#define MG_ENABLE_CUSTOM_MILLIS  0

#define MG_ENABLE_MBEDTLS  0

#define MG_ARCH MG_ARCH_CMSIS_RTOS2

#define MG_STMPACK_NET 0

#define MG_ENABLE_CUSTOM_RANDOM  1

// Translate to Mongoose macros
#if MG_STMPACK_NET == 0
#define MG_ENABLE_TCPIP 1
#elif MG_STMPACK_NET == 1
#define MG_ENABLE_LWIP 1
#elif MG_STMPACK_NET == 2
#define MG_ENABLE_FREERTOS_TCP 1
#elif MG_STMPACK_NET == 3
#define MG_ENABLE_RL 1
#endif

#if MG_ENABLE_PACKED_FS
#define MG_ENABLE_FILE 0
#endif

#endif
```

**Relevant pieces of code:**

**// Main task executing**
```
void StartTask02(void *argument)
{
  /* USER CODE BEGIN StartTask02 */
	struct mg_mgr mgr;

    MG_INFO(("Waiting for IP..."));


	//mg_log_set(MG_LL_DEBUG);

	mg_mgr_init(&mgr);


	struct mg_tcpip_driver_stm32_data driver_data = {.mdc_cr = 4};
	struct mg_tcpip_if mif = {.mac = GENERATE_LOCALLY_ADMINISTERED_MAC(),
	  	                      .driver = &mg_tcpip_driver_stm32,
	  	                      .driver_data = &driver_data,
							  .ip = 0x6420380A, //10.56.32.100
							  .mask = 0x000000FF,
							  .gw = 0x0101000A
	  						 };

	mg_tcpip_init(&mgr, &mif);


	mg_timer_add(&mgr, 3000, MG_TIMER_REPEAT | MG_TIMER_RUN_NOW, timer_fn, &mgr);
	  //mg_timer_add(&mgr, 3000, MG_TIMER_REPEAT | MG_TIMER_RUN_NOW, timer_fn, &mgr);
	  	//mg_listen(&mgr, "tcp://0.0.0.0:1234", cb, &mgr);
  /* Infinite loop */
	for(;;)
	{
		mg_mgr_poll(&mgr, 1000);
		HAL_GPIO_TogglePin(GPIOB, GPIO_PIN_14);
		osDelay(200);
	}
	mg_mgr_free(&mgr);
  /* USER CODE END StartTask02 */
}
```

**//Other functions/Callbacks**

```
void mg_random(void *buf, size_t len)
{
  extern RNG_HandleTypeDef hrng;
  for (size_t n = 0; n < len; n += sizeof(uint32_t))
  {
    uint32_t r;
    HAL_RNG_GenerateRandomNumber(&hrng, &r);
    memcpy((char *) buf + n, &r, n + sizeof(r) > len ? len - n : sizeof(r));
  }
}


static void fn(struct mg_connection *c, int ev, void *ev_data, void *fn_data)
{
  if (ev == MG_EV_OPEN)
  {
	  MG_INFO(("%lu CREATED", c->id));
  }
  if (ev == MG_EV_ERROR)
  {
	  // On error, log error message
	  MG_ERROR(("%lu ERROR %s", c->id, (char *) ev_data));
  }
  if (ev == MG_EV_CONNECT)
  {
	  // If target URL is SSL/TLS, command client connection to use TLS
	  if (mg_url_is_ssl(s_url))
	  {
		  struct mg_tls_opts opts = {.ca = "ca.pem"};
		  mg_tls_init(c, &opts);
      }
  }

  if (ev == MG_EV_MQTT_OPEN)
  {
    // MQTT connect is successful
    struct mg_str subt = mg_str(s_sub_topic);
    struct mg_str pubt = mg_str(s_pub_topic), data = mg_str("hello");
    struct mg_mqtt_opts sub_opts;

    //Log Connect Message
    MG_INFO(("%lu CONNECTED to %s", c->id, s_url));

    memset(&sub_opts, 0, sizeof(sub_opts));
    sub_opts.topic = subt;
    sub_opts.qos = s_qos;
    mg_mqtt_sub(c, &sub_opts);
    MG_INFO(("%lu SUBSCRIBED to %.*s", c->id, (int) subt.len, subt.ptr));
    struct mg_mqtt_opts pub_opts;
    memset(&pub_opts, 0, sizeof(pub_opts));
    pub_opts.topic = pubt;
    pub_opts.message = data;
    pub_opts.qos = s_qos, pub_opts.retain = false;
    mg_mqtt_pub(c, &pub_opts);
    MG_INFO(("%lu PUBLISHED %.*s -> %.*s", c->id, (int) data.len, data.ptr, (int) pubt.len, pubt.ptr));
  }

  if(ev == MG_EV_MQTT_MSG)
  {
	  // When we get echo response, print it
	  struct mg_mqtt_message *mm = (struct mg_mqtt_message *) ev_data;
	  MG_INFO(("%lu RECEIVED %.*s <- %.*s", c->id, (int) mm->data.len, mm->data.ptr, (int) mm->topic.len, mm->topic.ptr));
  }
  if (ev == MG_EV_CLOSE)
  {
	  MG_INFO(("%lu CLOSED", c->id));
	  s_conn = NULL;  // Mark that we're closed
  }
  (void) fn_data;
}


static void timer_fn(void *arg)
{
  struct mg_mgr *mgr = (struct mg_mgr *) arg;
  struct mg_mqtt_opts opts = {.clean = true,
                              .qos = s_qos,
                              .topic = mg_str(s_pub_topic),
                              .version = 4,
                              .message = mg_str("bye")};

  if (s_conn == NULL) s_conn = mg_mqtt_connect(mgr, s_url, &opts, fn, NULL);
}
```

Thank you for your attention and for your work!
Best Regards!

**Comment by scaprile** [comment, marked as answer]
TL;DR
Is this a discussion ? Looks like several issues scrambled together.
If you have issues, please file issues; be concise. short, precise, describe your scenario and your issue, keep focused on one issue, please.
What runs on what ? Where is Mongoose running ? What is your scenario ? Do you have a DHCP server ? What are your network addresses ? Did you read the documentation ? Did you see we claim the F767 has been tested or we wrote a driver for it anywhere ? (Because we did not) You mention you are using our built-in stack, then lwIP, then Windows, then Linux, try to keep your focus; some ports of lwIP in Cube and Keil are broken and we can't do anything about that.
You are having networking problems, did you happen to think on using a network sniffer ?
Please open an issue for each problem, follow the issue template, answer all questions
